### PR TITLE
Restore WebUI v2 always approval affordance

### PR DESCRIPTION
**Pull request by danielwpz**
## Summary
- add `allow_always` to WebUI v2 gate prompt/projection payloads and preserve it through live gate events and projection replay
- only advertise Always Allow for typed product-workflow approval gates; generic approval gates remain one-shot only
- allow persistent approval policies for manifest `Ask` capabilities while still rejecting `Deny`
- let profile approval replay accept same-user persistent approval grants, not only one-shot approval leases
- add targeted coverage for Ask-capability persistent approval eligibility, persistent grant reuse, and generic approval gate projection

## Validation
- `cargo fmt`
- `cargo fmt --check` earlier in this hotfix pass before the final review-comment commit
- `cargo clippy -p ironclaw_reborn_cli --features webui-v2-beta --bin ironclaw-reborn -- -D warnings`
- manual WebUI verification on local server: `Approve and always` wrote the persistent policy, later `builtin.http` and `builtin.write_file` calls replayed the persistent grant without a second approval gate

Note: no unit/integration test suite was run in this final pass; kept to fmt + lint per hotfix urgency.

### CODE REVIEW

File: crates/ironclaw_reborn_composition/src/slack_delivery.rs [hunk 1]
```diff
@@ -248,6 +248,7 @@ impl SlackFinalReplyDeliveryObserver {
                         headline: "Approval needed".to_string(),
                         body: "A step in the workflow requires your approval to resume."
                             .to_string(),
+                        allow_always: true,
```
Comment (serrrfirat) [inline crates/ironclaw_reborn_composition/src/slack_delivery.rs]: **🔴 Bug/Security (High, confidence 97):** `allow_always: true` is hardcoded for every `BlockedApproval`, while `turn_events.rs` correctly calls `is_approval_gate_ref(gate_ref)` to restrict to typed approval gates only.

Generic gates (e.g. `gate:approve-slack`) will get `allow_always: true` here but `persistent_approval_allowed()` will return `AlwaysAllowUnsupported` — breaking the Slack approval flow.

```rust
// Fix:
allow_always: ironclaw_product_workflow::is_approval_gate_ref(gate_ref),
```

File: crates/ironclaw_reborn_composition/src/profile_approval_authorization.rs [hunk 2]
```diff
@@ -164,10 +164,13 @@ fn has_matching_one_shot_approval_grant(
     let expected_grantee = Principal::Extension(context.extension_id.clone());
     let expected_user_approver = Principal::User(context.user_id.clone());
     context.grants.grants.iter().any(|grant| {
-        grant.capability == descriptor.id
-            && grant.constraints.max_invocations == Some(1)
+        let one_shot_approval_grant = grant.constraints.max_invocations == Some(1)
             && (grant.issued_by == Principal::HostRuntime
-                || grant.issued_by == expected_user_approver)
+                || grant.issued_by == expected_user_approver);
```
Comment (serrrfirat) [inline crates/ironclaw_reborn_composition/src/profile_approval_authorization.rs]: **🟡 Security/TOCTOU (Medium, confidence 78):** The `persistent_approval_grant` branch does not check `expires_at`. `active_grant()` filters expiry at load time but the resulting `CapabilityGrant` is frozen into `ExecutionContext.grants`. A grant that expires mid-turn still satisfies this check for all subsequent dispatches.

```rust
let persistent_approval_grant = grant.constraints.max_invocations.is_none()
    && grant.issued_by == expected_user_approver
    && grant.constraints.expires_at
           .map_or(true, |exp| exp > chrono::Utc::now());
```

File: crates/ironclaw_reborn_composition/src/profile_approval_authorization.rs [hunk 3]
```diff
@@ -164,10 +164,13 @@ fn has_matching_one_shot_approval_grant(
...
+        let persistent_approval_grant = grant.constraints.max_invocations.is_none()
+            && grant.issued_by == expected_user_approver;
```
Comment (serrrfirat) [inline crates/ironclaw_reborn_composition/src/profile_approval_authorization.rs]: **🟡 Bug (Medium, confidence 78):** `max_invocations.is_none() && issued_by == expected_user_approver` is not structurally unique to the persistent-approval flow. Any future standing user grant with the same shape (e.g. a bootstrap grant) would silently satisfy an approval gate. Consider a distinguishing marker — a dedicated issuer variant or a `GrantConstraints` flag — so only grants written by the persistent approval store match this branch.

File: crates/ironclaw_approvals/src/policy.rs [hunk 4]
```diff
@@ -22,7 +22,7 @@ const POLICY_PATH_CACHE_MAX_ENTRIES: usize = 1024;
 const POLICY_CAS_RETRY_ATTEMPTS: usize = 3;
 
 pub fn permission_mode_allows_persistent_approval(permission: PermissionMode) -> bool {
-    matches!(permission, PermissionMode::Allow)
+    matches!(permission, PermissionMode::Allow | PermissionMode::Ask)
```
Comment (serrrfirat) [inline crates/ironclaw_approvals/src/policy.rs:32]: **🟡 Security (Medium, confidence 70):** This expands persistent approval eligibility to `PermissionMode::Ask`. Extension developers relying on `default_permission = ask` to enforce mandatory per-invocation consent can no longer do so — `Ask` is now effectively a UX hint, not a security control.

Suggestion: add a doc-comment on `PermissionMode::Ask` or this function noting the semantics, so extension authors know they need a different mechanism for hard per-invocation gates.

File: crates/ironclaw_reborn_composition/src/projection/turn_events.rs [hunk 5]
```diff
@@ -330,6 +331,7 @@ async fn blocked_prompt_payload(
     let Some(gate_ref) = state.gate_ref.as_ref() else {
         return Ok(None);
     };
+    let allow_persistent_approval = is_approval_gate_ref(gate_ref);
```
Comment (serrrfirat) [inline crates/ironclaw_reborn_composition/src/projection/turn_events.rs]: **🟡 Conventions (Medium, confidence 88):** Per `.claude/rules/testing.md`: when a predicate gates a user-visible side effect through wrappers, a unit test on the predicate alone is insufficient.

`is_approval_gate_ref → blocked_prompt_payload → GatePromptView::allow_always` has no caller-level test for the positive typed-gate case. `webui_event_stream_does_not_offer_always_for_generic_approval_gate` covers the `false` path; add a matching test for the `true` path (`gate:approval-<uuid>` → `allow_always: true`).

**Nit (same line):** `is_approval_gate_ref` is computed unconditionally before the `match` but is only used in `BlockedApproval`. Move it inside that arm to avoid dead computation and misleading code for new `Blocked*` variants:
```rust
TurnStatus::BlockedApproval => Ok(Some(gate_prompt(
    event, gate_ref_str, "Approval required",
    is_approval_gate_ref(gate_ref),
))),
```

**Review by serrrfirat — COMMENTED** [review]
## Code Review — PR #4717: Restore WebUI v2 always approval affordance

**Reviewers:** Security · Bugs · Performance · Tests · Conventions
**Findings:** 3 High · 5 Medium · 3 Nit (11 total, confidence ≥ 50)

---

### 🔴 High

**[BUG/SECURITY] Slack delivery hardcodes `allow_always:true`, bypassing `is_approval_gate_ref` gate**
`crates/ironclaw_reborn_composition/src/slack_delivery.rs:251` | Also flagged by: Security (78), Conventions (95)

`turn_events.rs` correctly calls `is_approval_gate_ref(gate_ref)` to restrict `allow_always` to typed product-workflow approval gates. `slack_delivery.rs` hardcodes `allow_always: true` for every `BlockedApproval`, including generic gates. When a user clicks "Always Allow" on a generic gate in Slack, `persistent_approval_allowed()` returns `AlwaysAllowUnsupported` — broken UX and divergent security surface.

Fix: `allow_always: ironclaw_product_workflow::is_approval_gate_ref(gate_ref)`

---

**[TEST] Stale integration test will fail: `default_runtime_does_not_reuse_persistent_policy_for_manifest_ask`**
`crates/ironclaw_host_runtime/tests/host_runtime_persistent_approvals_contract.rs:352`

Test asserts `RuntimeFailureKind::Authorization` for an `PermissionMode::Ask` capability with a stored persistent policy. The PR changes `permission_mode_allows_persistent_approval` to return `true` for `Ask`, so `production.rs:981` now injects the stored grant into context. Capability dispatches successfully — test assertion fails in CI.

Fix: flip assertion to `assert!(matches!(outcome, RuntimeCapabilityOutcome::Dispatched(..)))` and rename to `default_runtime_reuses_persistent_policy_for_manifest_ask`. Add separate test for `PermissionMode::Deny` still blocking.

---

**[TEST] Slack `allow_always` behavior is untested — divergence from WebUI path undetected**
`crates/ironclaw_reborn_composition/src/slack_delivery.rs:245-260`

No test asserts the `allow_always` field value in Slack gate prompt output, so the unconditional `true` vs WebUI's `is_approval_gate_ref`-gated value is silently wrong.

Fix: Add `tests::slack_delivery::slack_approval_gate_with_typed_ref_sets_allow_always` and `slack_approval_gate_with_generic_ref_clears_allow_always`.

---

### 🟡 Medium

**[SECURITY/TOCTOU] No `expires_at` check for persistent grants in `has_matching_approval_grant`**
`crates/ironclaw_reborn_composition/src/profile_approval_authorization.rs:166-172` | Also flagged by: Security (72)

`persistent_approval_grant` branch accepts grants where `max_invocations.is_none()` without checking `expires_at`. `active_grant()` filters expiry at load time, but the grant is frozen into `ExecutionContext.grants` for the turn's lifetime. An expired persistent grant remains valid for all subsequent dispatches within the same long-running turn.

Fix: add `&& grant.constraints.expires_at.map_or(true, |exp| exp > chrono::Utc::now())`

---

**[SECURITY] `PermissionMode::Ask` persistent approval semantics reversal — undocumented**
`crates/ironclaw_approvals/src/policy.rs:24-26`

Extension developers who set `default_permission = ask` to enforce mandatory per-invocation consent can no longer rely on that mode to prevent standing approval grants. The change is intentional per PR body, but the `PermissionMode::Ask` documentation doesn't reflect the new semantics.

---

**[BUG] `persistent_approval_grant` predicate too broad — matches any user-issued `None`-invocation grant**
`crates/ironclaw_reborn_composition/src/profile_approval_authorization.rs:164-177`

`max_invocations.is_none() && issued_by == expected_user_approver` is not structurally unique to the persistent approval flow. Any future standing user grant with the same shape could silently satisfy an approval gate and bypass required confirmation.

---

**[TEST] Missing test: different-user persistent grant (`max_invocations=None`) must not satisfy gate**
`crates/ironclaw_reborn_composition/src/profile_approval_authorization.rs:141-171` | Also flagged by: Security (85)

`other_user_issued_approval_grant_does_not_allow_resume` only exercises the one-shot branch (`max_invocations: Some(1)`). The new persistent path's `issued_by == expected_user_approver` guard has no dedicated regression test.

Fix: add `tests::other_user_issued_persistent_grant_does_not_allow_resume` with `max_invocations: None` from a different user → `Decision::RequireApproval`.

---

**[TEST] `permission_mode_allows_persistent_approval` has no direct unit tests**
`crates/ironclaw_approvals/src/policy.rs:24-26`

The expanded function lacks direct tests for any variant. All coverage is indirect through multiple abstraction layers.

Fix: add `tests::policy::permission_mode_allows_persistent_approval_variants` — `Allow → true`, `Ask → true`, `Deny → false`.

---

**[CONVENTIONS] No caller-level test: `is_approval_gate_ref → blocked_prompt_payload → allow_always`**
`crates/ironclaw_reborn_composition/src/projection/turn_events.rs:300-395`
Anchor: `.claude/rules/testing.md` — when a predicate gates a side effect through wrappers, unit test on the helper alone is insufficient.

No integration-tier test drives `blocked_prompt_payload` with a typed approval gate ref and asserts `allow_always: true` in the resulting `GatePromptView`. The generic-gate test (`webui_event_stream_does_not_offer_always_for_generic_approval_gate`) covers only `allow_always: false`.

---

### 🟢 Nit

- `is_approval_gate_ref` computed unconditionally before `match event.status` but used only in `BlockedApproval` arm — move inside the arm (`turn_events.rs:334`)
- `ProductProjectionItem::Gate` with `allow_always` has no serde round-trip test; adding one would lock in the `#[serde(default)]` backward-compat contract (`outbound.rs:655-662`)
- `is_approval_gate_ref` has no isolation unit tests; only covered indirectly via integration tests (`gate_ref.rs:12`)

**Review by sunglow666 — DISMISSED** [review]
Reviewed the approval policy/runtime path, projection payloads, WebUI v2 gate normalization/replay path, and schema/adapters touched by this PR. No blocking findings.

Validation run locally:
- `cargo fmt --all -- --check`
- `git diff --check main...HEAD`
- `node --test crates/ironclaw_webui_v2_static/static/js/pages/chat/lib/gates.test.mjs crates/ironclaw_webui_v2_static/static/js/pages/chat/lib/useChatEvents.test.mjs`
- `cargo test -p ironclaw_webui_v2 --test webui_v2_schema_contract`
- `cargo test -p ironclaw_reborn_composition webui_event_stream`
- `cargo test -p ironclaw_reborn_composition user_issued_persistent_approval_grant_allows_reuse`
- `cargo test -p ironclaw_reborn_composition active_extension_capability`

Notes: I did not run the full Rust suite. The PR's `NODE_OPTIONS=--experimental-default-type=module ...` form is not supported by my local Node, but the `.mjs` tests pass with plain `node --test`.

**Review by sunglow666 — DISMISSED** [review]
Re-reviewed the latest head (`9e0d3e200`) after the follow-up approval feedback. No blocking findings.

The new changes address the prior review concerns I checked:

- `PermissionMode::Ask` persistent-approval semantics are documented and covered directly.
- Persistent approval reuse now requires the dedicated persistent-approval issuer and rejects user-issued persistent-like grants, other-user grants, and expired grants.
- WebUI/event projection restoration preserves `allow_always` for typed approval gates while generic gates stay false.
- Slack delivery no longer hardcodes `allow_always: true`; it follows typed approval gate refs.
- Projection payload serde covers `allow_always` roundtrip and legacy missing-field defaulting.
- Host runtime manifest-ask persistent approval contract is updated and passing.

Validation run locally:

- `cargo fmt --all -- --check`
- `git diff --check main...HEAD`
- `node --test crates/ironclaw_webui_v2_static/static/js/pages/chat/lib/gates.test.mjs crates/ironclaw_webui_v2_static/static/js/pages/chat/lib/useChatEvents.test.mjs`
- `cargo test -p ironclaw_approvals persistent_approval`
- `cargo test -p ironclaw_product_workflow is_approval_gate_ref_accepts_only_typed_approval_prefix`
- `cargo test -p ironclaw_reborn_composition persistent_approval`
- `cargo test -p ironclaw_reborn_composition persistent_like`
- `cargo test -p ironclaw_reborn_composition --features slack-v2-host-beta slack_approval_prompt`
- `cargo test -p ironclaw_product_adapters --features test-support projection_state`
- `cargo test -p ironclaw_reborn_composition webui_event_stream_offers_always_for_typed_approval_gate`
- `cargo test -p ironclaw_webui_v2 --test webui_v2_schema_contract`
- `cargo test -p ironclaw_host_runtime default_runtime_reuses_persistent_policy_for_manifest_ask`

Notes: I did not run the full Rust suite. The plain `cargo test -p ironclaw_product_adapters projection_state` invocation requires `test-support` for existing integration-test fakes, so I reran it with that feature. `ironclaw_reborn_composition` still emits two existing test-build warnings in `runtime.rs`, unrelated to this PR path.

**Comment by danielwpz** [comment]
Addressed the latest review feedback and resolved the handled inline threads.

Changes pushed:
- `9e0d3e200` fixes the review findings:
  - Slack approval prompts now gate `allow_always` with `is_approval_gate_ref`, so generic Slack approval gates stay one-shot.
  - Persistent approval replay now requires a dedicated persistent-approval issuer marker instead of accepting any user-issued unlimited grant.
  - Profile approval replay re-checks `expires_at` before accepting one-shot or persistent approval grants.
  - `PermissionMode::Ask` persistent approval semantics are documented.
  - Added coverage for Slack typed/generic `allow_always`, persistent-grant marker/expiry/other-user cases, direct permission-mode behavior, WebUI typed approval projection, projection serde compatibility, and `is_approval_gate_ref`.
  - Updated the manifest-`ask` persistent approval runtime contract to match the intended behavior.
- `2bf3d59f9` fixes the Code Style no-panic failure by removing the production `.expect()` from the persistent approval issuer helper.

Local validation after the final fix:
- `cargo fmt`
- `python3 scripts/check_no_panics.py --base a09b241df6bbfa5681d5825dd32afc18cfc1c859 --head HEAD`
- `cargo clippy -p ironclaw_reborn_cli --features webui-v2-beta --bin ironclaw-reborn -- -D warnings`

Current CI status observed on the latest head: Code Style, Tests (Reborn), Reborn E2E, and Regression Test Check are passing; Tests (Legacy) is still in progress.

**Comment by danielwpz** [comment]
Follow-up: latest head `2bf3d59f9` is now green across the CI runs I was tracking:

- Code Style: success
- Tests (Reborn): success
- Reborn E2E: success
- Regression Test Check: success
- Tests (Legacy): success

No unresolved review threads are open at this point.
